src/service/async_service_impl.cc

```diff
@@ -0,0 +1,168 @@
+#include "async_service_impl.h"
+#include "src/config/get_config.h"
+#include <boost/asio.hpp>
+#include <boost/thread/thread.hpp>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/server_builder.h>
+
+namespace authservice {
+namespace service {
+
+class ServiceState {
+public:
+  virtual ~ServiceState() = default;
+
+  virtual void Proceed() = 0;
+};
+
+class ProcessingState;
+
+class CompleteState : public ServiceState {
+public:
+  explicit CompleteState(ProcessingState *processor) : processor_(processor) {
+  }
+
+  void Proceed() override;
+
+private:
+  ProcessingState *processor_;
+};
+
+class ProcessingState : public ServiceState {
+public:
+  ProcessingState(std::shared_ptr<authservice::config::Config> config, Authorization::AsyncService &service,
+                  grpc::ServerCompletionQueue &cq, boost::asio::io_context &io_context)
+          : service_(service), cq_(cq), responder_(&ctx_), io_context_(io_context), config_(std::move(config)),
+            impl_(*config_) {
+    spdlog::trace("Creating processor state");
+    service.RequestCheck(&ctx_, &request_, &responder_, &cq_, &cq_, this);
+  }
+
+  void Proceed() override {
+    // Spawn a new instance to serve new clients while we process this one
+    new ProcessingState(config_, service_, cq_, io_context_);
```

Comment:
Shouldn't this class hold a pointer to this object from the heap and delete it in its destructor? Maybe doesn't matter too much in practice because we know in practice that this is called by `main`. Still technically a memory leak though.  :)

Reply:
This creates a new instance to handle any next request after the current one. We can't delete the state of the next request in the destructor of the current request. It's not really a leak because this handler for the next request is placed on the completion queue and will be later pulled off by `cq_->Next()` then processed and ultimately deleted by the destructor of `CompleteState`. This is the same way the [async gRPC example](https://github.com/grpc/grpc/blob/v1.24.0/examples/cpp/helloworld/greeter_async_server.cc) works.

Personally I don't really like using raw `new` like this either, but the way that async gRPC and its completion queue work makes it very difficult to avoid.

I'll try to add a few more comments to explain exactly what's happening here.